nexus/reconfigurator/planning/src/planner.rs

```diff
@@ -1561,29 +1559,60 @@ impl<'a> Planner<'a> {
         // Of the out-of-date zones, filter out zones that can't be updated yet,
         // either because they're not ready or because it wouldn't be safe to
         // bounce them.
-        let mut updateable_zones = Vec::new();
-        for (sled_id, zone, new_image_source) in out_of_date_zones {
-            if self.are_zones_ready_for_updates(mgs_updates)
-                && self.can_zone_be_updated(&zone, &mut report)?
-            {
-                updateable_zones.push((sled_id, zone, new_image_source));
-            }
+        let (nexus_updateable_zones, non_nexus_updateable_zones): (
+            Vec<_>,
+            Vec<_>,
+        ) = out_of_date_zones
+            .into_iter()
+            .filter(|(_, zone, _)| {
+                self.are_zones_ready_for_updates(mgs_updates)
+                    && self.can_zone_be_shut_down_safely(&zone, &mut report)
+            })
+            .partition(|(_, zone, _)| zone.zone_type.is_nexus());
```

Comment:
Take it or leave it: I'm wondering if we can make `should_nexus_zone_be_expunged()` more type-safe (so it can't panic on bad input) with something like (this is untested):

```suggestion
        let (nexus_updateable_zones, non_nexus_updateable_zones): (
            Vec<_>,
            Vec<_>,
        ) = out_of_date_zones
            .into_iter()
            .filter(|(_, zone, _)| {
                self.are_zones_ready_for_updates(mgs_updates)
                    && self.can_zone_be_shut_down_safely(&zone, &mut report)
            })
            .map(|(sled_id, zone, image_source)| {
                let nexus_config = match &zone.z_type {
                    blueprint_zone_type::Nexus(nexus_config) => nexus_config,
                    _ => None,
                };
                (sled_id, zone, image_source, nexus_config)
            })
            .partition(|(_, _, _, nexus_config)| nexus_config.is_none());
```

Then we'll already have the `nexus_config` and can pass it to `should_nexus_zone_be_expunged()`.  Not a big deal.

Reply:
I would be more on-board for this, but I think the type of `nexus_config` would be wrapped in an option here, which would need to be unwrapped anyway?